Add trivy scan daily build

neil-hickey opened the pull request:
- Done as part of https://github.com/vmware-tanzu/carvel-kapp-controller/issues/368
- Add slack notification to triyy scanner

Signed-off-by: Neil Hickey <nhickey@vmware.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

If this is your first time, please read our contributor guidelines: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/CONTRIBUTING.md and developer guide https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/docs/dev.md
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
Fixes #368

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. 

-->
```release-note
NONE
```

#### Additional Notes for your reviewer:
The final step is to add a secret to github that contains the slack bot with write access to the slack channel.

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

neil-hickey committed:
Add trivy scan for docker image

- Done as part of https://github.com/vmware-tanzu/carvel-kapp-controller/issues/368
- Add slack notification to triyy scanner

Signed-off-by: Neil Hickey <nhickey@vmware.com>

neil-hickey commented:
For line: https://github.com/vmware-tanzu/carvel-kapp-controller/pull/424/files#diff-d2f2b520bb7bfb36f97c6ec62a237288585d8f9d4301d9b31cb0c3dfad98f1c9R74

I think we can re-use the same bot that the kapp-team are using (given it contains the correct access to our slack instance) - I don't have the ability to add this secret / copy the secret from  https://github.com/vmware-tanzu/carvel-kapp which I think will be the simplest move to get this working.

danielhelfand commented:
`Test_PackageInstalled_FromPackageInstall_DeletionFailureBlocks` is what [failed in CI](https://github.com/vmware-tanzu/carvel-kapp-controller/runs/4217966568?check_suite_focus=true). This test has been flaky, and we should look into it outside of this pr.

danielhelfand commented on .github/workflows/trivy-scan.yml:
```diff
@@ -0,0 +1,77 @@
+        uses: actions/setup-go@v1
+        with:
+          go-version: "1.17"
+      - name: Install Carvel Tools
+        uses: vmware-tanzu/carvel-setup-action@v1
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          only: ytt, kapp, kbld, vendir
+          ytt: v0.37.0
+          kapp: v0.42.0
+          kbld: v0.31.0
+          vendir: v0.23.0
+      - name: Build image from Dockerfile
```
I don't think we need this, right? Looks like all that is happening here is a `docker build` and trivy-related things. So Carvel tools shouldn't be installed.

neil-hickey replied:
Right, I realized I am missing a commit which tests the kapp-controller binary in addition to the docker image. That's why I had it in there. I am going to pull this PR back to draft while I add that in and then re-open it up

neil-hickey committed:
Change to use triby source instead of action

- The feedback cycle is too long
- It is too complicated to show a summary in slack with docker image
  file type from trivy.
- Therefore I opted to display a link which brings you to the relevant
  tab in github when new vulnerabilites are found

Signed-off-by: Neil Hickey <nhickey@vmware.com>

danielhelfand commented on .github/workflows/trivy-scan.yml:
```diff
@@ -0,0 +1,88 @@
+
+          # download the sarif format template
+          git clone --depth 1 https://github.com/aquasecurity/trivy
+      - name: Run Trivy Reports
```
Can any of this be captured in the trivy config so there is less scripting?

neil-hickey replied:
Hmm, I am not sure! Its a good shout, maybe we can reduce this down with some config file. 👍

neil-hickey replied:
Trivy doesn't have any sort of config / defaults. However they do have some ENV_VARs you can use. I made a new commit that extracts some of the duplication to variables. I think it makes it a fair bit clear 👍

danielhelfand commented on .github/workflows/trivy-scan.yml:
```diff
@@ -0,0 +1,88 @@
+          brew install aquasecurity/trivy/trivy
+
+          # download the sarif format template
+          git clone --depth 1 https://github.com/aquasecurity/trivy
```
How is this used? Is there a different way of getting this than a git clone?

neil-hickey replied:
So this is used by the following line:

`--format template --template "@trivy/contrib/sarif.tpl"`

The sarif.tpl is a template file that exists in the trivy repository, I couldn't find another way to get it. It's not built into the trivy scanner binary itself, so this seemed simplest

danielhelfand replied:
Should we maybe check this file into the kapp-controller repo and update regularly? Not sure if we want to pull directly from main branch of trivy to get this.

neil-hickey replied:
It felt cheaper to me for trivy folks to look after that template rather than us. Maybe we start with this, and if it ends up breaking a lot or they remove it, we can pull it into the repo? I am hesitant to take on the template ourselves, given a bunch of the carvel tools will be using the same template eventually and we would have to duplicate it into each repo. @danielhelfand

danielhelfand replied:
Sounds good to me.

cppforlife commented on .github/workflows/trivy-scan.yml:
```diff
@@ -0,0 +1,88 @@
+          ./hack/build.sh
+
+          # docker image
+          docker build -t docker.io/carvel/kapp-controller:${{ github.sha }} .
```
hmm... it seems that we are building this image from scratch instead of analyzing last release image, which i think would mean that we would always be getting latest updates to images -- so no cves caught.

neil-hickey replied:
The outcome from our conversation RE this comment:

We will create another daily scan that gets the latest released docker image and scans it.

neil-hickey replied:
This ended up being more complicated, getting the scan results into the security tag based on a tag doesn't seem to work for me. So I am going to move this out into another issue. I think having a daily run against HEAD is good enough that I don't want to block on previously released scanning.

neil-hickey committed:
Move trivy flags to env vars

- This removes some of the complexity involved in lots of flags.
- Where a flag is not used, trivy will ignore it.

Signed-off-by: Neil Hickey <nhickey@vmware.com>
